### PR TITLE
Stl algo

### PR DESCRIPTION
**Pull request by MoustaphaSaadSymbyo**

### CODE REVIEW

File: cpprelude/include/cpprelude/iterator.h [hunk 1]
```diff
@@ -124,11 +124,12 @@ namespace cpprelude
 	}
 
 	template<typename T>
-	struct sequential_iterator
-	{
+	struct sequential_iterator:std::iterator<std::random_access_iterator_tag, T, isize, T*, T&>
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: mmmmm ... since you use some of the definitions from the std::iterator i think it's better to do the integration by adding the typedefs directly instead of inheritance.

Comment (MoustaphaSaadSymbyo) [reply cpprelude/include/cpprelude/iterator.h]: and i think by eliminating `std::iterator` this should eliminate the need of `#include <iterator>`

File: cpprelude/include/cpprelude/iterator.h [hunk 2]
```diff
@@ -2,7 +2,7 @@
 
 #include "cpprelude/defines.h"
 #include "cpprelude/memory.h"
-
+#include<iterator>
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: leave a space between `#include` and `<file>` for visual aid

File: cpprelude/include/cpprelude/iterator.h [hunk 3]
```diff
@@ -152,6 +154,22 @@ namespace cpprelude
 			return result;
 		}
 
+	
+		sequential_iterator<T>&
+		operator+=(typename base::difference_type offset)
+		{
+			_element += offset;
+			return *this;	
+
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: remove unnecessary whitespaces

File: cpprelude/include/cpprelude/iterator.h [hunk 4]
```diff
@@ -152,6 +154,22 @@ namespace cpprelude
...
+		}
+
+		sequential_iterator<T>
+		operator+(typename base::difference_type offset)const
+		{
+			sequential_iterator<T> temp = *this;
+			return temp += offset;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: `return _element + offset;`

File: cpprelude/include/cpprelude/iterator.h [hunk 5]
```diff
@@ -167,6 +185,25 @@ namespace cpprelude
 			return result;
 		}
 
+		sequential_iterator<T>&
+		operator-=(typename base::difference_type offset)
+		{
+			return this += -offset;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: don't see the benefit of using the definition of `operator+=` in the implementation of all the operators.
since they are too trivial to implement

File: cpprelude/include/cpprelude/iterator.h [hunk 6]
```diff
@@ -167,6 +185,25 @@ namespace cpprelude
...
+		}
+
+		sequential_iterator<T>
+		operator-(typename base::difference_type offset)
+		{
+			sequential_iterator<T> temp = *this;
+			return temp += -offset;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: `return _element - offset;`

File: cpprelude/include/cpprelude/iterator.h [hunk 7]
```diff
@@ -179,6 +216,30 @@ namespace cpprelude
 			return !operator==(other);
 		}
 
+		bool
+		operator<(const sequential_iterator<T>& other) const
+		{
+			return other - *this > 0;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: wait what?!
why not `_element < other._element`?!

File: cpprelude/include/cpprelude/iterator.h [hunk 8]
```diff
@@ -179,6 +216,30 @@ namespace cpprelude
...
+		}
+
+		bool
+		operator>(const sequential_iterator<T>& other) const
+		{
+			return other < *this ;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: why?!

File: cpprelude/include/cpprelude/iterator.h [hunk 9]
```diff
@@ -179,6 +216,30 @@ namespace cpprelude
...
+		}
+
+		bool 
+		operator<=(const sequential_iterator<T>& other) const
+		{
+			return !(other > *this);
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: why?!

File: cpprelude/include/cpprelude/iterator.h [hunk 10]
```diff
@@ -179,6 +216,30 @@ namespace cpprelude
...
+		}
+
+		bool
+		operator>=(const sequential_iterator<T>& other) const
+		{
+			return !(other < *this);
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: why?!

File: cpprelude/include/cpprelude/iterator.h [hunk 11]
```diff
@@ -212,13 +273,20 @@ namespace cpprelude
 		{
 			return _element;
 		}
+
+		T&
+		operator[](typename base::difference_type offset)
+		{
+			return *(*this += offset);
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: why?!
`return _element[offset];`

File: cpprelude/include/cpprelude/iterator.h [hunk 12]
```diff
@@ -212,13 +273,20 @@ namespace cpprelude
...
+		}
+
 	};
 
 	template<typename T>
 	struct const_forward_iterator;
 
 	template<typename T>
-	struct forward_iterator
+	struct forward_iterator:std::iterator<std::forward_iterator_tag, T, isize, T*, T&>
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: same thing about std::iterator

File: cpprelude/include/cpprelude/iterator.h [hunk 13]
```diff
@@ -298,7 +366,7 @@ namespace cpprelude
 	};
 
 	template<typename T>
-	struct const_forward_iterator
+	struct const_forward_iterator:std::iterator<std::forward_iterator_tag, const T, isize,const T*, const T&>
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: same thing about std::iterator

File: cpprelude/include/cpprelude/iterator.h [hunk 14]
```diff
@@ -373,7 +441,7 @@ namespace cpprelude
 	struct const_bidirectional_iterator;
 
 	template<typename T>
-	struct bidirectional_iterator
+	struct bidirectional_iterator:std::iterator<std::bidirectional_iterator_tag, T, isize, T*, T&>
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: same thing about std::iterator

File: cpprelude/include/cpprelude/iterator.h [hunk 15]
```diff
@@ -475,7 +543,7 @@ namespace cpprelude
 	};
 
 	template<typename T>
-	struct const_bidirectional_iterator
+	struct const_bidirectional_iterator:std::iterator<std::bidirectional_iterator_tag, const T, isize, const T*, const T&>
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: same thing about std::iterator

File: cpprelude/include/cpprelude/iterator.h [hunk 16]
```diff
@@ -572,12 +640,15 @@ namespace cpprelude
 	struct const_bucket_array_iterator;
 
 	template<typename T, usize bucket_size>
-	struct bucket_array_iterator
+	struct bucket_array_iterator:std::iterator<std::random_access_iterator_tag, T, isize, T*, T&>
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: same thing about std::iterator

File: cpprelude/include/cpprelude/iterator.h [hunk 17]
```diff
@@ -628,6 +699,34 @@ namespace cpprelude
...
+			}
+			
+			return *this;
+		}
+
+		bucket_array_iterator
+		operator+(typename base::difference_type offset)const
+		{
+			auto temp = *this;
+			return temp += offset;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: this is some what complex so either do as you did here or create a private function -private means a function that has an _ before it's name- for example `_move(usize)` that moves the iterator either forward or backward

File: cpprelude/include/cpprelude/iterator.h [hunk 18]
```diff
@@ -572,12 +640,15 @@ namespace cpprelude
...
 	template<typename T, usize bucket_size>
-	struct bucket_array_iterator
+	struct bucket_array_iterator:std::iterator<std::random_access_iterator_tag, T, isize, T*, T&>
 	{
+		using base = std::iterator<std::random_access_iterator_tag, T, isize, T*, T&>;
 		using data_type = T;
 		T** _bucket_it;
 		T* _element_it;
 		usize _index;
+		usize _bucket_index = 0;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: why these two new member variables?

File: cpprelude/include/cpprelude/iterator.h [hunk 19]
```diff
@@ -628,6 +699,34 @@ namespace cpprelude
 			return result;
 		}
 
+		bucket_array_iterator&
+		operator+=(typename base::difference_type offset)
+		{
+			typename base::difference_type _offset = offset + _index;
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: just name it offset or maybe work using the same variable as the parameter

File: cpprelude/include/cpprelude/iterator.h [hunk 20]
```diff
@@ -1241,6 +1528,20 @@ namespace cpprelude
 		{
 			return !operator==(other);
 		}
+		/*
+		bool
```
Comment (MoustaphaSaadSymbyo) [inline cpprelude/include/cpprelude/iterator.h]: remove commented code

File: scratch/src/main.cpp [hunk 21]
```diff
@@ -4,5 +4,6 @@ int
 main(int argc, char** argv)
 {
 	do_benchmark();
+
```
Comment (MoustaphaSaadSymbyo) [inline scratch/src/main.cpp]: revert this change

File: unittest/src/unittest_bucket_array.cpp [hunk 22]
```diff
@@ -216,7 +216,7 @@ TEST_CASE("bucket_array test", "[bucket_array]")
 		while (count--)
 		{
 			CHECK(*it == i++);
-			it = next(it);
+			it = cpprelude::next(it);
```
Comment (MoustaphaSaadSymbyo) [inline unittest/src/unittest_bucket_array.cpp]: why?!

**Review by MoustaphaSaadSymbyo — COMMENTED** [review]
please add unit tests for the sorting functions

**Comment by MoustaphaSaadSymbyo** [comment]
please add unit tests for the sorting functions

File: cpprelude/include/cpprelude/iterator.h (hunk 18 above)
Comment (hanaa-mohamed) [reply cpprelude/include/cpprelude/iterator.h]: debugging variables and I forgot to remove them. sorry, bad me.

File: cpprelude/include/cpprelude/iterator.h (hunk 17 above)
Comment (hanaa-mohamed) [reply cpprelude/include/cpprelude/iterator.h]: 👍

File: unittest/src/unittest_bucket_array.cpp (hunk 22 above)
Comment (hanaa-mohamed) [reply unittest/src/unittest_bucket_array.cpp]: without specifying the namespace, it had some compilation errors due to ambiguity at resolving which namespace to use (std or cpprelude) .

File: cpprelude/include/cpprelude/iterator.h (hunk 8 above)
Comment (hanaa-mohamed) [reply cpprelude/include/cpprelude/iterator.h]: ok, some mistake happened here. nice you spotted it.

File: cpprelude/include/cpprelude/iterator.h (hunk 9 above)
Comment (hanaa-mohamed) [reply cpprelude/include/cpprelude/iterator.h]: ok, some mistake happened here. nice you spotted it.
